cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,549 @@
+        return acceptReplicaSwap(srcReplicaToSwap, dstReplicaToSwap);
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  private ActionAcceptance acceptReplicaSwap(Replica srcReplicaToSwap, Replica dstReplicaToSwap) {
+    if (!srcReplicaToSwap.isLeader() && !dstReplicaToSwap.isLeader()) {
+      // Swapping two follower replicas is always accepted
+      return ACCEPT;
+    }
+    String srcTopicName = srcReplicaToSwap.topicPartition().topic();
+    String dstTopicName = dstReplicaToSwap.topicPartition().topic();
+    if (srcReplicaToSwap.isLeader() && dstReplicaToSwap.isLeader() && Objects.equals(srcTopicName, dstTopicName)) {
+      // Swapping two leader replicas of the same topic is always accepted
+      return ACCEPT;
+    }
+    if (doesLeaderReplicaRemoveViolateOptimizedGoal(srcReplicaToSwap) || doesLeaderReplicaRemoveViolateOptimizedGoal(dstReplicaToSwap)) {
+      return REPLICA_REJECT;
+    }
+    return ACCEPT;
+  }
+
+  /**
+   * Check whether this goal is violated after removing the given leader replica. It is violated if there is not be enough
+   * leader replicas on that broker for that topic
+   *
+   * @param replicaToBeRemoved replica to be removed
+   * @return {@code true} if the given replica move would violate the optimized goal (i.e. the move is not acceptable),
+   * {@code false} otherwise.
+   */
+  private boolean doesLeaderReplicaRemoveViolateOptimizedGoal(Replica replicaToBeRemoved) {
+    if (!replicaToBeRemoved.isLeader()) {
+      // Moving a follower replica does not violate/affect this goal
+      return false;
+    }
+    String topicName = replicaToBeRemoved.topicPartition().topic();
+    if (!_mustHaveLeaderReplicaPerBrokerTopics.contains(topicName)) {
+      // Moving a replica of a irrelevant topic does not violate/affect this goal
+      return false;
+    }
+    int topicLeaderReplicaCountOnSourceBroker = replicaToBeRemoved.broker().numLeaderReplicasOfTopicInBroker(topicName);
+    return topicLeaderReplicaCountOnSourceBroker <= minTopicLeadersPerBroker();
+  }
+
+  /**
+   * This is a hard goal; hence, the proposals are not limited to dead broker replicas in case of self-healing.
+   * Sanity Check: The total number of leader replicas of certain topics is sufficient so that these leader replicas
+   * can be moved to satisfy this goal.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    if (_balancingConstraint.topicsWithMinLeadersPerBrokerPattern().pattern().isEmpty()) {
+      _mustHaveLeaderReplicaPerBrokerTopics = Collections.emptySet();
+      return;
+    }
```

Comment:
(This is a followup for the earlier comment https://github.com/linkedin/cruise-control/pull/1429#discussion_r567204294)

Can we just call the following function first, then use `_mustHaveLeaderReplicaPerBrokerTopics` in a following if statement to see if it is empty?
```
_mustHaveLeaderReplicaPerBrokerTopics = Collections.unmodifiableSet(Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel::topics));
```

This 
1. eliminates code duplication and
2. ensures that if `_mustHaveLeaderReplicaPerBrokerTopics` empty despite having a non-empty pattern, we terminate early.

Am I missing something?

Reply:
Good idea